core/src/main/java/io/grpc/internal/AbstractReferenceCounted.java

```diff
@@ -29,44 +29,50 @@
  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
-package io.grpc.okhttp;
+package io.grpc.internal;
 
-import com.google.common.base.Preconditions;
-
-import com.squareup.okhttp.ConnectionSpec;
-
-import io.grpc.internal.ClientTransport;
-import io.grpc.internal.ClientTransportFactory;
-
-import java.util.concurrent.ExecutorService;
-
-import javax.net.ssl.SSLSocketFactory;
+import java.util.concurrent.atomic.AtomicInteger;
 
 /**
- * Factory that manufactures instances of {@link OkHttpClientTransport}.
+ * Base implementation for all reference counted objects.
  */
-class OkHttpClientTransportFactory implements ClientTransportFactory {
-  private final String host;
-  private final int port;
-  private final ExecutorService executor;
-  private final String authorityHost;
-  private final SSLSocketFactory sslSocketFactory;
-  private final ConnectionSpec connectionSpec;
+public abstract class AbstractReferenceCounted implements ReferenceCounted {
+  private final AtomicInteger refCount = new AtomicInteger(1);
+
+  @Override
+  public final int referenceCount() {
+    return refCount.get();
+  }
 
-  public OkHttpClientTransportFactory(String host, int port, String authorityHost,
-      ExecutorService executor, SSLSocketFactory factory, ConnectionSpec connectionSpec) {
-    this.host = Preconditions.checkNotNull(host, "host");
-    this.port = port;
-    this.executor = Preconditions.checkNotNull(executor, "executor");
-    this.authorityHost = Preconditions.checkNotNull(authorityHost, "authorityHost");
-    this.sslSocketFactory = factory;
-    this.connectionSpec = connectionSpec;
+  @Override
+  public final ReferenceCounted retain() {
+    int newCount = refCount.incrementAndGet();
+    if (newCount <= 0) {
```

Comment:
`<= 1`, or use `getAndIncrement()`. We always want to notice if we bring a dead object back to life. It'd be _nice_ if the increment didn't happen if `referenceCount()` was <= 0 upon entry, but I can understand not dealing with the whole loop-until-it-works model that that would require.